### ISSUE TITLE
With several Chromium-based browsers, adding one to favorites also add others

### ISSUE THREAD

**Issue by DynamicDonkey** [falzonv/discreet-launcher#66]
Hi there! Been using this launcher for a while now and I love it! However one of the last couple updates introduced a bug in the favorites system. Adding Brave to the favorites tab also adds Chrome. Opening the manage favorites dialog and removing Chrome doesn't remove it either.

If there's any more info I can provide, let me know and I will. Thanks!

**Comment by falzonv** [comment]
Hello,

Thank you very much for your feedback, I am glad you like the launcher!

Quite surprising bug indeed, which version of the launcher do you use currently? (can be found at the end of the Settings)

It would be useful to know which version you had before the bug appeared but you may not have noticed it.
Do you know approximately since how many time you face this bug? (might give a clue about the used version at this time)

Could you make an export (Settings > Export/Import > Export) and share here the lines starting with "favorites.txt:" which are related to Brave and/or Chrome?
*(you can also attach the complete text file, as you prefer, what I would like to see is how it is stored in the favorites when you add Brave and that Chrome is automatically added)*

---

Other hypothesis:
Apparently Brave and Chrome are both based on Chromium, I wonder if this could be related.
Did you update one of them when you started to notice the bug?
If so, the version before/after would be useful (for me to check what they changed), but probably even more difficult to retrieve as they have a release rhythm much more intense than me...

Best regards.

**Comment by DynamicDonkey** [comment]
Launcher version 3.6.0 from F-Droid. Seems they haven't gotten the 4.0 and 4.0.1 yet.

As for `favorites.txt:`, there are two lines that are `com.google.android.apps.chrome.Main`. Looks like a Brave update has caused this bug.

The entirety of export is as follows, just in case.
```
# Export Discreet Launcher v3.6.0 (Jun 10, 2021 2:03:05 PM)
# Be careful: if you edit this file, it may produce unexpected results!
#
# Internal text files content
favorites.txt: com.google.android.apps.chrome.Main
favorites.txt: com.google.android.apps.chrome.Main
favorites.txt: org.thoughtcrime.securesms.RoutingActivity
favorites.txt: com.snapchat.android.LandingPageActivity
favorites.txt: com.spotify.music.MainActivity
hidden.txt: com.vincent_falzon.discreetlauncher.ActivityMain
#
shortcuts.txt: none
shortcuts_legacy.txt: none
# Settings
display_clock: false
clock_format: HH:mm
transparent_status_bar: false
force_portrait: false
immersive_mode: false
reverse_interface: false
hide_app_names: false
icon_pack: none
notification: false
#
# Saved icons
#
```
I'm not sure what version of Brave caused this bug, as I kind of just started to deal with it and there's been a couple updates since it first occurred. Current version I have installed of Brave is 1.24.86, released on May 19th of this year.

**Comment by falzonv** [comment]
Hello,

Thank you very much for these information, I have found the root cause and it is on my side.
After further investigation, this bug is present since v1.2.0 (from 16th March) and affects all Chromium-based browsers :-/

**Explanation**

- At the beginning (v1.0.0), I was using the package name (ex: "com.brave.browser") to distinguish applications from each other because this is something which must be unique on the stores.
- However, a user told me that several applications could come from the same package (#3, with a camera and gallery apps), so it was not really unique anymore and I had to find something more precise.
- There was the name of the main activity of the application: it started with the package name and continued with the application entry point (for example if we had a "com.gallerycamera" package, it could contain "com.gallerycamera.gallery.Main" and "com.gallerycamera.camera.Main") => I decided to use this and it solved the issue.
- But at the same time, it introduced this quite unexpected bug and I discovered the reason with the result of your export and the source code of an icon pack: all the Chromium-based browsers use the exact same name for the entry point and it doesn't include their package name (I didn't know it was even possible!).

**Solution**

It is of course to store both the package name and the entry point name in the favorites file, but I have to find a way to do this without breaking the current favorites selections (to convert from the current format to the new one without the users losing all their favorites settings, and also managing imports).

Unfortunately I will not have my computer this week-end and will not be able to start working on solving this before at least next Monday in the evening. I am sorry for this :-/

Best regards.

**Comment by DynamicDonkey** [comment]
This is awesome how quickly you figured this out, thank you for the support! Kind of strange they would all use the same path to the main activity, but Chromium is a large project with tons of quirks.

Take all the time you need. You've more than blessed the FOSS community with this wonderful project!

**Comment by falzonv** [comment]
Hello @DynamicDonkey,

Thank you very much for the kind words :-)

I am pleased to inform you that this has been solved in new release v4.1.0!
This release has been published a few minutes ago.

Here are the usual availability delays for each location:
- 4 to 8 days on F-Droid
- 0 to 2 days on Google Play
- Already available as direct APK on GitHub (https://github.com/falzonv/discreet-launcher/releases)

To be as transparent as possible for users, the internal files are automatically converted to the new format. It should work flawlessly for 99% applications, but may produce strange results for the Chromium-based browsers if the user has several. The reason is that the conversion function looks for the current name in the internal file (where they are all the same for Chromium), looks for this name in the applications list and rewrite the package name + the activity path in the internal file (so it may pick the wrong one).
=> **Conclusion: after this automatic conversion is done, you may have to adjust manually for Chromium-based browsers.**

*Remarks:
I will leave this ticket opened until the release is effectively published on F-Droid.
You can use the export/import feature to switch from one version to another.*

Best regards.

**Comment by falzonv** [comment]
Since Thursday, release v4.1.0 is available on F-Droid so I will close this ticket.
Feel free to add comments if needed or wished.
